keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java

```diff
@@ -99,6 +165,231 @@ private static INDArray broadcastMinus(INDArray a, INDArray b) {
+        if (mask.length() == 1 || right.length() == 1) {
+            long[] resultShape = Shape.broadcastOutputShape(mask.shape(), right.shape());
+            if (mask.length() == 1) {
+                mask = Nd4j.valueArrayOf(right.shape(), mask.getDouble(0));
+                Nd4j.getExecutioner().exec(
+                    new OldLessThan(mask, right, mask, mask.length())
+                );
+            } else {
+                Nd4j.getExecutioner().exec(
+                    new OldLessThan(mask,
+                        valueArrayOf(mask.shape(), right.getDouble(0), bufferType),
+                        mask,
+                        mask.length()
+                    )
+                );
+            }
+            return mask.reshape(resultShape);
+        }
+        Nd4j.getExecutioner().exec(
+            new OldLessThan(mask, right, mask, mask.length())
+        );
+        return mask;
+    }
+
+    public static INDArray getLessThanOrEqualToMask(INDArray mask, INDArray right, DataBuffer.Type bufferType) {
+        if (mask.length() == 1 || right.length() == 1) {
+            long[] resultShape = Shape.broadcastOutputShape(mask.shape(), right.shape());
+            if (mask.length() == 1) {
+                mask = Nd4j.valueArrayOf(right.shape(), mask.getDouble(0));
+                Nd4j.getExecutioner().exec(
+                    new OldLessThanOrEqual(mask, right, mask, mask.length())
+                );
+            } else {
+                Nd4j.getExecutioner().exec(
+                    new OldLessThanOrEqual(mask,
+                        valueArrayOf(mask.shape(), right.getDouble(0), bufferType),
+                        mask,
+                        mask.length()
+                    )
+                );
+            }
+            return mask.reshape(resultShape);
+        }
+        Nd4j.getExecutioner().exec(
+            new OldLessThanOrEqual(mask, right, mask, mask.length())
+        );
+        return mask;
+    }
+
+    private static INDArray performOperationWithScalarTensorPreservingShape(INDArray left, INDArray right, BiFunction<INDArray, INDArray, INDArray> operation) {
+        long[] resultShape = Shape.broadcastOutputShape(left.shape(), right.shape());
+        INDArray result;
+        if (left.length() == 1) {
+            result = operation.apply(Nd4j.valueArrayOf(right.shape(), left.getDouble(0)), right);
+        } else {
+            result = operation.apply(left, Nd4j.valueArrayOf(left.shape(), right.getDouble(0)));
+        }
+        return result.reshape(resultShape);
+    }
+
+    private static boolean shapeAIsSmallerThanShapeB(long[] shapeA, long[] shapeB) {
+        if (shapeA.length != shapeB.length) {
+            return shapeA.length < shapeB.length;
+        }
+        for (int ind = 0; ind < shapeA.length; ind++) {
+            if (shapeA[ind] < shapeB[ind]) {
+                return true;
```

Comment:
so {1, 100} is smaller than {2, 1}, for example? Yet {1, 100} is larger than {1000}. It's a strange definition.